### PR TITLE
[llvm] Add KnownBits implementations for avgFloor and avgCeil

### PR DESCRIPTION
**Pull request by changkhothuychung**
This PR is to address the issue #84640

### CODE REVIEW

**Comment by changkhothuychung** [comment]
@RKSimon RkSim
What type of test should I add, should it be similar to this test ? 

https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/unittests/Support/KnownBitsTest.cpp#L834-L848

**Comment by changkhothuychung** [comment]
@RKSimon Rkso
just following up on this!

**Comment by goldsteinn** [comment]
> @RKSimon RkSim What type of test should I add, should it be similar to this test ?
> 
> https://github.com/llvm/llvm-project/blob/ffe41819e58365dfbe85a22556c0d9d284e746b9/llvm/unittests/Support/KnownBitsTest.cpp#L834-L848

Your unit tests should look something like:
```
  testBinaryOpExhaustive(KnownBits::abds, APIntOps::abds,
                         checkCorrectnessOnlyBinary);
```

The first argument is a function pointer to the knownbits function we are testing.
The second argument is a function pointer to the reference implementation on an `APInt`.
The third is optional. If your implementation is complete (i.e proves all knowable bits) you can
drop it, otherwise just use what you see above. We like to strive for exhaustive, so if yours is not,
can you provide a few values that is fails for and explain a bit why its not really feasible?

File: llvm/lib/Support/KnownBits.cpp
Comment (goldsteinn) [inline llvm/lib/Support/KnownBits.cpp:773]: Why `computeForSatAddSub` as opposed to just `computeForAddSub`?
The setup implies `nsw` so I don't think you will ever sat... (likewise below, the U versions all imply `nuw` and the S versions all imply `nsw`)

File: llvm/include/llvm/Support/KnownBits.h
Comment (jayfoad) [inline llvm/include/llvm/Support/KnownBits.h:357]: Putting the bit twidling formula in the comment is not helpful. Suggestion:
```suggestion
  /// Compute knownbits resulting from APIntOps::avgFloorS
```
Same for the others.

**Review by RKSimon — CHANGES_REQUESTED** [review]
#86754 has now landed so please can you update the SelectionDAG::computeKnownBits AVG cases to use KnowBits implementations

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3424]: you don't need these opcode checks anymore since you have split the cases. Likewise below.

File: llvm/lib/Support/KnownBits.cpp
Comment (goldsteinn) [inline llvm/lib/Support/KnownBits.cpp:799]: still missing nsw/nuw flags.

Comment (changkhothuychung) [reply llvm/lib/Support/KnownBits.cpp]: oh sorry I missed this one. By the way, can you explain to me what Sat, nsw and nuw mean for these functions? I dont really understand what they are and how they work. Thanks!

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
Comment (jayfoad) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:3428]: The zext/extract is stuff is unnecessary and wrong.
```suggestion
    Known = KnownBits::avgFloorU(Known, Known2);
```

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (RKSimon) [inline llvm/unittests/Support/KnownBitsTest.cpp:570]: You shouldn't need lambdas for these:
```cpp
testBinaryOpExhaustive(KnownBits::avgFloorS, APIntOps::avgFloorS, checkCorrectnessOnlyBinary);
```

Comment (goldsteinn) [reply llvm/unittests/Support/KnownBitsTest.cpp:570]: What values are there not complete for?

File: llvm/lib/Support/KnownBits.cpp
Comment (goldsteinn) [reply llvm/lib/Support/KnownBits.cpp]: `nuw` means that when we do `X + Y` it won't have unsigned overflow. So `X nuw+ Y >= X && X nuw+ Y >= Y` must hold. `nsw` is the same but for the signbit.
So with 8bit, `128 + 128` violates both nuw/nsw.

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (jayfoad) [reply llvm/unittests/Support/KnownBitsTest.cpp:570]: I think you could probably make the KnownBits implementation optimal by using the ext/add/shift/trunc method.

**Review by RKSimon — CHANGES_REQUESTED** [review]
Please address the KnownBitsTests failures

**Comment by RKSimon** [comment]
Any luck with fixing the unit test failures?

**Comment by changkhothuychung** [comment]
> Any luck with fixing the unit test failures?

no not yet unfortunately, it should be related to the lines in the file SelectionDAG.cpp I highlighted above, be cause the failed test links to those lines, but currently I still struggle figuring out why it fails.

**Comment by goldsteinn** [comment]
> > Any luck with fixing the unit test failures?
> 
> no not yet unfortunately, it should be related to the lines in the file SelectionDAG.cpp I highlighted above, be cause the failed test links to those lines, but currently I still struggle figuring out why it fails.

Just some tips for reproducing locally.
For the known bits unittests you can do:
```
$> ninja UnitTests && unittests/Support/./SupportTests --gtest_filter=KnownBitsTest*
```

For the DAGCombiner tests you can do:
```
$> ninja && ./bin/llvm-lit ../llvm/test/CodeGen/
```

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (goldsteinn) [inline llvm/unittests/Support/KnownBitsTest.cpp:573]: Which cases are not optimal?

**Comment by RKSimon** [comment]
@changkhothuychung If you change the KnownBits implementation back to the same as the original SelectionDAG (add/sub of ext) do the problems go away?

**Comment by changkhothuychung** [comment]
> @changkhothuychung If you change the KnownBits implementation back to the same as the original SelectionDAG (add/sub of ext) do the problems go away?

Hi, no it would not. However, I just found there might be a potential bug related to the previous PR https://github.com/llvm/llvm-project/pull/86754. 

In the file [AArch64SelectionDAGTest.cpp](https://github.com/llvm/llvm-project/pull/86754/files#diff-4977843fe788410d80ba3e9bf106458157803452f0fc91d2a88cb5df170daa9f), at line 828, it's 

```cpp 

SDValue AVGFLOORS =
      DAG->getNode(ISD::AVGFLOORU, Loc, Int16Vec8VT, ZextOp0, ZextOp1);
      
 ```
 
 I think ISD::AVGFLOORS should be passed as the first argument instead. But the tests still passed.

File: llvm/unittests/CodeGen/AArch64SelectionDAGTest.cpp
Comment (RKSimon) [inline llvm/unittests/CodeGen/AArch64SelectionDAGTest.cpp:829]: I thought I'd fixed this diff in 6acc50ae5b5971af563a0f40b5f0279870171541 ? Please can you rebase

**Review by RKSimon — COMMENTED** [review]
Only 1 minor typo outstanding - thanks

**Comment by changkhothuychung** [comment]
> Only 1 minor typo outstanding - thanks

Let me know when you have a chance to review, after that I will merge the PR. Thanks!

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:787]: Put this in a helper function like `KnownBits avg(const KnownBits &LHS, const KnownBits &RHS, bool IsCeil, bool IsSigned)`? It is weird to have this code duplicated four times.

Comment (goldsteinn) [reply llvm/lib/Support/KnownBits.cpp:787]: Especially when it is generically implemented just with hardset `IsCeil` and `IsSign`.

File: llvm/include/llvm/Support/KnownBits.h
Comment (jayfoad) [inline llvm/include/llvm/Support/KnownBits.h:370]: This helper function should not be in the header. It can just be a static helper file in `KnownBits.cpp`.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:21]: Don't reformat code that you haven't changed.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:49]: Don't reformat code that you haven't changed.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:779]: It might be neater to pass LHS and RHS in by value, instead of by const ref, so that you don't need to copy them here.

File: llvm/unittests/Support/KnownBitsTest.cpp
Comment (jayfoad) [inline llvm/unittests/Support/KnownBitsTest.cpp:506]: I don't think you need `false` here because the knownbits implementation should be optimal.

**Review by jayfoad — COMMENTED** [review]
This looks good to me now except for the remaining cosmetic issues.

File: llvm/lib/Support/KnownBits.cpp
Comment (goldsteinn) [reply llvm/lib/Support/KnownBits.cpp:802]: hmm?

**Comment by goldsteinn** [comment]
This LGTM, wait one 1 more approval to push.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [reply llvm/lib/Support/KnownBits.cpp:21]: This still looks like spurious reformatting.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [reply llvm/lib/Support/KnownBits.cpp:49]: This still looks like spurious reformatting.

File: llvm/lib/Support/KnownBits.cpp
Comment (jayfoad) [inline llvm/lib/Support/KnownBits.cpp:478]: This looks like spurious reformatting.

**Comment by changkhothuychung** [comment]
@jayfoad f
looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?

**Comment by jayfoad** [comment]
> looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?

Oh, I thought this implementation was supposed to be optimal, but maybe I got that wrong.

Patch LGTM anyway.

**Review by jayfoad — APPROVED** [review]

**Comment by changkhothuychung** [comment]
> > looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?
> 
> Oh, I thought this implementation was supposed to be optimal, but maybe I got that wrong.
> 
> Patch LGTM anyway.

I am assuming I need to put back the flag to disable optimality check ?

**Comment by goldsteinn** [comment]
> > > looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?
> > 
> > 
> > Oh, I thought this implementation was supposed to be optimal, but maybe I got that wrong.
> > Patch LGTM anyway.
> 
> I am assuming I need to put back the flag to disable optimality check ?

You already do. Its the `bool` argument.

**Comment by jayfoad** [comment]
> > > looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?
> > 
> > Oh, I thought this implementation was supposed to be optimal, but maybe I got that wrong.
> > 
> > Patch LGTM anyway.
> 
> I am assuming I need to put back the flag to disable optimality check ? 

Yes (at least for the signed functions CeilS and FloorS) because you can't commit with failing tests.

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by changkhothuychung** [comment]
> > > > looks like the BinaryExhaustive test fails when I remove the bool flag to enable Optimal check?
> > > 
> > > 
> > > Oh, I thought this implementation was supposed to be optimal, but maybe I got that wrong.
> > > Patch LGTM anyway.
> > 
> > 
> > I am assuming I need to put back the flag to disable optimality check ?
> 
> You already do. Its the `bool` argument.

I made a push to put it back after I made that comment.

**Comment by changkhothuychung** [comment]
Thanks everyon! looks like I got 2 approves already, I will merge this PR shortly if there is no other concern.

**Comment by changkhothuychung** [comment]
@RKSimon RK
it looks like there is a failed build in later stages after the PR was merged - https://lab.llvm.org/buildbot/#/builders/185/builds/6813

Should we revert the changes?

**Comment by RKSimon** [comment]
wait until the current build has finished to see if it was an intermittent fail

**Pull request merged**
